server/apps/subscriptions.go

```diff
@@ -4,31 +4,149 @@
 }
 
 type subscriptions struct {
 	configurator configurator.Service
+	mm           *pluginapi.Client
 }
 
 var _ Subscriptions = (*subscriptions)(nil)
 
-func NewSubscriptions(configurator configurator.Service) Subscriptions {
+func NewSubscriptions(mm *pluginapi.Client, configurator configurator.Service) Subscriptions {
 	return &subscriptions{
+		mm:           mm,
 		configurator: configurator,
 	}
 }
 
-func (subs *subscriptions) GetSubscriptionsForChannel(subj SubscriptionSubject, channelID string) ([]*Subscription, error) {
-	return []*Subscription{
-		{
-			AppID:     "Hello",
-			Subject:   SubjectUserJoinedChannel,
-			ChannelID: "",
-		},
-	}, nil
+// GetSubscriptionsForChannelOrTeam returns subscriptions for a given subject and
+// channelID or teamID from the store
+func (subs *subscriptions) GetSubscriptionsForChannelOrTeam(subj SubscriptionSubject, channelOrTeamID string) ([]*Subscription, error) {
+	key := GetSubsKVkey(subj, channelOrTeamID)
+	var savedSubs []*Subscription
+	if err := subs.mm.KV.Get(key, &savedSubs); err != nil {
+		return nil, errors.Wrap(err, "failed to get saved subscriptions")
+	}
+
+	return savedSubs, nil
+}
+
+func (subs *subscriptions) GetSubscriptionsForApp(app string, subj SubscriptionSubject, channelID string) ([]*Subscription, error) {
+	// if subj is nil, grab all subjects for the
+	return nil, nil
+}
+
+// StoreSubscription stores a subscription for a change notification
+// TODO move this to store package or file
+func (subs *subscriptions) StoreSubscription(subj SubscriptionSubject, sub Subscription, channelID string) error {
+	key := GetSubsKVkey(subj, channelID)
+
+	// get all subscriptions for the subject
+	var savedSubs []*Subscription
+	if err := subs.mm.KV.Get(key, &savedSubs); err != nil {
+		return errors.Wrap(err, "failed to get saved subscriptions")
+	}
+
+	// check if sub exists
+	var newSubs []*Subscription
+	foundSub := 0
+	for _, s := range savedSubs {
+		// modify the sub to the latest request
+		if s.SubscriptionID == sub.SubscriptionID {
+			foundSub++
+			newSubs = append(newSubs, &sub)
+			continue
+		}
+		newSubs = append(newSubs, s)
+	}
+	if foundSub == 0 {
+		newSubs = append(newSubs, &sub)
+	}
+
+	// sub exists. update and save updated subs
+	_, err := subs.mm.KV.Set(key, newSubs)
+	if err != nil {
+		return errors.Wrap(err, "failed to save subscriptions")
+	}
+	return nil
+}
+
+// DeleteSubscription deletes a subscription
+func (subs *subscriptions) DeleteSubscription(subj SubscriptionSubject, subID SubscriptionID, channelID string) error {
+	key := GetSubsKVkey(subj, channelID)
+
+	// get all subscriptions for the subject
+	var savedSubs []*Subscription
+	if err := subs.mm.KV.Get(key, &savedSubs); err != nil {
+		return errors.Wrap(err, "failed to get saved subscriptions")
+	}
+
+	// check if sub exists
+	var newSubs []*Subscription
+	for _, s := range savedSubs {
+		if s.SubscriptionID == subID {
+			continue
+		}
+		newSubs = append(newSubs, s)
+	}
```

Comment:
```suggestion
	for i, s := range savedSubs {
		if s.SubscriptionID == subID {
			newSubs = append(newSubs, savedSubs[i+1:]...)
			break
		}
		newSubs = append(newSubs, s)
	}
```

What about something like this?

Reply:
That will work nicely and allow exiting the loop earlier, if possible.  Lev and I discussed building out a POC with minimal effort on efficiency concerns.  Thanks for this enhancement

Reply:
done.